### PR TITLE
Node syncing - cleanup the API and address review comments

### PR DESCRIPTION
**Pull request by adinapoli-mndc**
This PR addresses previous comments from @kim and @cloudhead. On top of that, it also cleans up the API considerably, and make the syncing code genuinely polymorphic.

After we merge this one, only two things will be missing (coming in one last final PR):

- [ ] Wire the syncing code to the protocol, at startup;
- [ ] Think about ways we can test the `syncNode` function (as that is essentially an endless loop);
- [ ] Telemetry

### CODE REVIEW

File: src/Oscoin/Protocol/Sync.hs [hunk 1]
```diff
@@ -176,35 +201,35 @@ data SyncContext c tx s m = SyncContext
     -- variation of it.
     , scLocalChainReader  :: BlockStoreReader c tx s m
     -- ^ A read-only handle to the local blockchain.
-    , scUpstreamConsumers :: [SyncEvent c tx s -> m ()]
+    , scEventHandlers :: [SyncEvent c tx s -> m ()]
     -- ^ A list of \"upstream consumers\", represented as effectful actions from
```
Comment (cloudhead) [inline src/Oscoin/Protocol/Sync.hs]: Comment needs updating :)

File: src/Oscoin/Protocol/Sync.hs [hunk 2]
```diff
@@ -1,9 +1,14 @@
+-- | Syncing nodes in oscoin
+--
+-- Based on the spec available at https://hackmd.io/2cPkrWTjTIWo2EmHZpPIQw
```
Comment (cloudhead) [inline src/Oscoin/Protocol/Sync.hs]: This is not permanent, I'd open a separate PR so we can merge it into the repo perhaps.

Comment (adinapoli-mndc) [reply src/Oscoin/Protocol/Sync.hs]: Yeah. I have been thinking about it, and I'd like to experiment with post-processing in hackmd so perhaps we can render all those nice graphs & latex into a `.pdf` (or similar). In the meantime I have opened an issue to track that (cfr. #597 ). Do you think we can tackle this separately?

File: src/Oscoin/Protocol/Sync.hs (hunk 1 above)
Comment (adinapoli-mndc) [reply src/Oscoin/Protocol/Sync.hs]: Can do here or in my (final) PR (which I will hopefully submit by the end of the day) -- whatever works best for you 😉

File: src/Oscoin/Protocol/Sync.hs (hunk 2 above)
Comment (cloudhead) [reply src/Oscoin/Protocol/Sync.hs]: Sure :+1:

File: src/Oscoin/Protocol/Sync.hs (hunk 1 above)
Comment (cloudhead) [reply src/Oscoin/Protocol/Sync.hs]: :man_shrugging:

**Pull request merged**